### ISSUE TITLE
CI test failure: Connection handshake was canceled due to the configured timeout

### ISSUE THREAD

**Issue by jkotas** [dotnet/runtime#99074]
```
    System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates [FAIL]
      Assert.Throws() Failure: Exception type was not an exact match
      Expected: typeof(System.Security.Authentication.AuthenticationException)
      Actual:   typeof(System.Net.Quic.QuicException)
      ---- System.Net.Quic.QuicException : Connection handshake was canceled due to the configured timeout of 00:00:10 seconds elapsing.
      -------- System.OperationCanceledException : The operation was canceled.
      Stack Trace:
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs(390,0): at System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates()
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs(397,0): at System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates()
        --- End of stack trace from previous location ---
        ----- Inner Stack Trace -----
        /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs(294,0): at System.Net.Quic.QuicListener.StartConnectionHandshake(QuicConnection connection, SslClientHelloInfo clientHello)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs(264,0): at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
        /_/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncTaskMethodBuilderT.cs(372,0): at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs(913,0): at System.Threading.ThreadPoolWorkQueue.Dispatch()
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.NonBrowser.cs(102,0): at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
        --- End of stack trace from previous location ---
        /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs(186,0): at System.Net.Quic.QuicListener.AcceptConnectionAsync(CancellationToken cancellationToken)
        /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs(390,0): at System.Net.Quic.Tests.MsQuicTests.<>c__DisplayClass8_0.<<CertificateCallbackThrowPropagates>b__2>d.MoveNext()
        --- End of stack trace from previous location ---
        ----- Inner Stack Trace -----
        /_/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs(119,0): at System.Net.Quic.ValueTaskSource.TryComplete(Exception exception)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs(1203,0): at System.Threading.CancellationTokenSource.CallbackNode.ExecuteCallback()
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs(810,0): at System.Threading.CancellationTokenSource.ExecuteCallbackHandlers(Boolean throwOnFirstException)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationTokenSource.cs(696,0): at System.Threading.CancellationTokenSource.NotifyCancellation(Boolean throwOnFirstException)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs(683,0): at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool)
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs(331,0): at System.Threading.TimerQueue.FireNextTimers()
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs(913,0): at System.Threading.ThreadPoolWorkQueue.Dispatch()
        /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.NonBrowser.cs(102,0): at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
        --- End of stack trace from previous location ---
        /_/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs(176,0): at System.Net.Quic.ValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
        /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs(241,0): at System.Net.Quic.QuicListener.StartConnectionHandshake(QuicConnection connection, SslClientHelloInfo clientHello)
      Output:
        System.Net.Quic.QuicException: User configured callback failed.
         ---> System.ArithmeticException: foobar
           at System.Net.Quic.Tests.MsQuicTests.<>c__DisplayClass8_0.<CertificateCallbackThrowPropagates>b__1(Object sender, X509Certificate cert, X509Chain chain, SslPolicyErrors errors) in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:line 383
           at System.Net.Quic.QuicConnection.SslConnectionOptions.ValidateCertificate(X509Certificate2 certificate, Span`1 certData, Span`1 chainData) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.SslConnectionOptions.cs:line 225
           --- End of inner exception stack trace ---
           at System.Net.Quic.QuicConnection.SslConnectionOptions.ValidateCertificate(X509Certificate2 certificate, Span`1 certData, Span`1 chainData) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.SslConnectionOptions.cs:line 252
           at System.Net.Quic.QuicConnection.SslConnectionOptions.ValidateCertificate(X509Certificate2 certificate, Span`1 certData, Span`1 chainData) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.SslConnectionOptions.cs:line 225
           at System.Net.Quic.QuicConnection.SslConnectionOptions.StartAsyncCertificateValidation(IntPtr certificatePtr, IntPtr chainPtr)
           at System.Net.Quic.ValueTaskSource.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ValueTaskSource.cs:line 176
           at System.Net.Quic.QuicConnection.FinishConnectAsync(QuicClientConnectionOptions options, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:line 365
           at System.Net.Quic.QuicConnection.<ConnectAsync>g__StartConnectAsync|2_0(QuicClientConnectionOptions options, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:line 79
           at System.Net.Quic.QuicConnection.<ConnectAsync>g__StartConnectAsync|2_0(QuicClientConnectionOptions options, CancellationToken cancellationToken) in /_/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs:line 94
           at System.Net.Quic.Tests.MsQuicTests.<>c__DisplayClass8_0.<<CertificateCallbackThrowPropagates>b__3>d.MoveNext() in /_/src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicTests.cs:line 388
        --- End of stack trace from previous location ---
           at Xunit.Assert.RecordExceptionAsync(Func`1 testCode) in /_/src/Microsoft.DotNet.XUnitAssert/src/Record.cs:line 105
MsQuic Counters:
    CONN_CREATED 1577
    CONN_HANDSHAKE_FAIL 25
    CONN_APP_REJECT 8
    CONN_LOAD_REJECT 0
```

## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=583082
Build error leg or test failing: System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates
Pull request: https://github.com/dotnet/runtime/pull/99033

## Error Message

Fill the error message using [step by step known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "System.Net.Quic.QuicException : Connection handshake was canceled due to the configured timeout of 00:00:10 seconds elapsing.",
  "ErrorPattern": "",
  "BuildRetry": false,
  "ExcludeConsoleLog": false
}
```

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=583082
**Error message validated:** `[System.Net.Quic.QuicException : Connection handshake was canceled due to the configured timeout of 00:00:10 seconds elapsing.`]
**Result validation:** :white_check_mark: Known issue matched with the provided build.
**Validation performed at:** 2/28/2024 7:50:19 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[584197](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584197)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584197&view=ms.vss-test-web.build-test-results-tab&runId=14058350&resultId=206302)|dotnet/runtime#99019|
|[584189](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584189)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=584189&view=ms.vss-test-web.build-test-results-tab&runId=14058174&resultId=206301)|dotnet/runtime#99048|
|[583901](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583901)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583901&view=ms.vss-test-web.build-test-results-tab&runId=14046748&resultId=206301)|dotnet/runtime#97529|
|[583884](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583884)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583884&view=ms.vss-test-web.build-test-results-tab&runId=14046626&resultId=206301)|dotnet/runtime#98908|
|[583812](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583812)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583812&view=ms.vss-test-web.build-test-results-tab&runId=14045038&resultId=206301)|dotnet/runtime#95565|
|[583795](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583795)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583795&view=ms.vss-test-web.build-test-results-tab&runId=14044162&resultId=206301)|dotnet/runtime#99094|
|[583790](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583790)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583790&view=ms.vss-test-web.build-test-results-tab&runId=14044172&resultId=206301)|dotnet/runtime#99092|
|[583448](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583448)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583448&view=ms.vss-test-web.build-test-results-tab&runId=14040576&resultId=206301)|dotnet/runtime#98117|
|[583418](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583418)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583418&view=ms.vss-test-web.build-test-results-tab&runId=14033164&resultId=206301)|dotnet/runtime#99076|
|[583402](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583402)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583402&view=ms.vss-test-web.build-test-results-tab&runId=14032936&resultId=206301)|dotnet/runtime#99073|
|[583430](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583430)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583430&view=ms.vss-test-web.build-test-results-tab&runId=14032590&resultId=206301)||
|[583376](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583376)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583376&view=ms.vss-test-web.build-test-results-tab&runId=14031510&resultId=206301)|dotnet/runtime#98958|
|[583380](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583380)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583380&view=ms.vss-test-web.build-test-results-tab&runId=14031416&resultId=206301)|dotnet/runtime#98958|
|[583295](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583295)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583295&view=ms.vss-test-web.build-test-results-tab&runId=14028892&resultId=206301)|dotnet/runtime#99059|
|[583262](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583262)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583262&view=ms.vss-test-web.build-test-results-tab&runId=14028414&resultId=206301)|dotnet/runtime#99025|
|[583200](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583200)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583200&view=ms.vss-test-web.build-test-results-tab&runId=14027674&resultId=206301)|dotnet/runtime#99024|
|[583183](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583183)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583183&view=ms.vss-test-web.build-test-results-tab&runId=14026832&resultId=206301)|dotnet/runtime#99058|
|[583154](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583154)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583154&view=ms.vss-test-web.build-test-results-tab&runId=14026236&resultId=206301)|dotnet/runtime#99042|
|[583106](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583106)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583106&view=ms.vss-test-web.build-test-results-tab&runId=14024580&resultId=206301)|dotnet/runtime#96941|
|[583082](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583082)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583082&view=ms.vss-test-web.build-test-results-tab&runId=14023696&resultId=206301)|dotnet/runtime#99033|
|[583094](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583094)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583094&view=ms.vss-test-web.build-test-results-tab&runId=14023496&resultId=215252)|dotnet/runtime#99053|
|[583054](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583054)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583054&view=ms.vss-test-web.build-test-results-tab&runId=14022996&resultId=206301)|dotnet/runtime#99049|
|[583031](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583031)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583031&view=ms.vss-test-web.build-test-results-tab&runId=14021458&resultId=206301)|dotnet/runtime#99051|
|[583028](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583028)|dotnet/runtime|[System.Net.Quic.Tests.MsQuicTests.CertificateCallbackThrowPropagates](https://dev.azure.com/dnceng-public/public/_build/results?buildId=583028&view=ms.vss-test-web.build-test-results-tab&runId=14021444&resultId=206301)|dotnet/runtime#98986|
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|24|24|24|

**Comment by ManickaP** [comment]
@rzikm do you think this might be related to your recent change in cert validation?

**Comment by rzikm** [comment]
It is possible let me have a look

**Comment by rzikm** [comment]
Investigation notes:
The failures seem to be concentrated on Alpine images which have MsQuic built from main, according to the commit hash of the libversion, the msquic library version should contain all the necessary changes for the feature to work.

Still curious why this failure didn't show up on the original PR

**Comment by ManickaP** [comment]
> Still curious why this failure didn't show up on the original PR

It's likely it didn't run on that set up at all. From our perspective, successful run of the test vs didn't run at all is really almost impossible to distinguish.

**Comment by rzikm** [comment]
Looks like https://github.com/microsoft/msquic/pull/4145 was not enough, We will have to disable the test for now, as this will take some time to fix.

**Comment by rzikm** [comment]
I think I figured out what is happening, and it's more complicated than I thought. According to the traces and packet captures I collected, following happens around the certificate validation

- We yield validation procedure to threadpool, freeing MsQuic worker thread
- MsQuic proceeds to send Handshake Acks, and waits for us to finish validation
- Custom certificate validation callback throws, we catch the exception and set it to `_connectedTcs`
https://github.com/dotnet/runtime/blob/56af1078adf3e179868a0e893d89c2ea2ca04836/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.SslConnectionOptions.cs#L138-L141
- Then we proceed to set the validation result in MsQuic
https://github.com/dotnet/runtime/blob/56af1078adf3e179868a0e893d89c2ea2ca04836/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.SslConnectionOptions.cs#L159-L162
- The above call, unfortunately, just queues the operation to a queue to be served by MsQuic worker.
- Moving on, the same .NET Threadpool thread then proceeds to process the continuation in ConnectAsync, this ValueTask was completed via `_connectedTcs`.
https://github.com/dotnet/runtime/blob/56af1078adf3e179868a0e893d89c2ea2ca04836/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L79
- Since the await rethrows the exception, control goes to catch block where the connection is disposed via DisposeAsync, which in turn calls MsQuicConnectionShutdown, which queues the shutdown **to the front of the queue**
- Eventually, MsQuic Worker will process the shutdown and prepares the connection for **application layer shutdown** with the DefaultConnectionCloseErrorCode specified by user.
- Since the application layer shutdown was processed first, connection close from rejecting the certificate gets ignored, and the logic I added in https://github.com/microsoft/msquic/pull/4145 does not trigger (it covered only **transport layer shutdown**).

**Comment by ManickaP** [comment]
@rzikm Could you also similarly describe what the problem is and how it manifests (not just the root cause)?

**Comment by rzikm** [comment]
One way to look at it saying this is a data race between `ConnectionCertificateValidationComplete` and `ConnectionShutdown` when called in a quick succession (due to how MsQuic queues the operations, this is a race even if we call them from the same thread in .NET)

**Comment by rzikm** [comment]
There's an underlying bug in MsQuic related to https://github.com/microsoft/msquic/issues/4132. MsQuic should not attempt to send application CONNECTION_CLOSE before handshake completes https://github.com/microsoft/msquic/issues/4166

> I understand this, I was more interested in how this translates to Handshake timeout on the other side.

Server is not allowed to process 1RTT packets before the handshake completes (for server, this means "before client accepts the certificate" by sending Finished message on TLS level). So the server receives 1RTT message with CONNECTION_CLOSE, but is waiting until handshake finishes so it is allowed to process it, and this times out

**Comment by rzikm** [comment]
This issue will be resolved in 2.4 release of MsQuic.

**Comment by liveans** [comment]
This issue is fixed with msquic 2.4 release but looks like we're still disabling this test against msquic versions which is lower than 2.4, I think we should decide about what to do with it.

- We can either convert `ActiveIssue` attribute to `Version check logic inside code` and let it run on all test suite and close this issue.
- We can run this `feature` with msquic versions which is higher than 2.4

I believe we can do the first option for the short term, but second option might be more reliable option here (has some perf downsides).

I think we were going to discuss it, right? @rzikm

**Comment by rzikm** [comment]
> I think we were going to discuss it, right? @rzikm

Last time I mentioned it in the team the feeling was to guarantee correct behavior even at cost of some perf.

Since MsQuic 2.4.1+ is available, I don't expect bumping the required version for the feature is going to adversely affect anyone.
